samples/routing/flowcontrol/Core_6/Shared/FlowControl/MarkerProcessor.cs

```diff
@@ -0,0 +1,68 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using NServiceBus;
+using NServiceBus.Features;
+using NServiceBus.Pipeline;
+using NServiceBus.Routing;
+using NServiceBus.Transports;
+
+class MarkerProcessor : ForkConnector<ITransportReceiveContext, IRoutingContext>
+{
+
+    public MarkerProcessor(EndpointInstance endpointInstance, int maxAckBatchSize)
+    {
+        this.maxAckBatchSize = maxAckBatchSize;
+        endpointName = endpointInstance.Endpoint.ToString();
+        instanceString = endpointInstance.ToString();
+    }
+
+    #region ProcessMarkers
+    public override async Task Invoke(ITransportReceiveContext context, Func<Task> next, Func<IRoutingContext, Task> fork)
+    {
+        await next().ConfigureAwait(false);
+
+        string sessionId;
+        string markerString;
+        string controlAddress;
+        string key;
+        if (!context.Message.Headers.TryGetValue("NServiceBus.FlowControl.Marker", out markerString) ||
+            !context.Message.Headers.TryGetValue("NServiceBus.FlowControl.Key", out key) ||
+            !context.Message.Headers.TryGetValue("NServiceBus.FlowControl.ControlAddress", out controlAddress) ||
+            !context.Message.Headers.TryGetValue("NServiceBus.FlowControl.SessionId", out sessionId))
+        {
+            return;
+        }
+
+        var marker = long.Parse(markerString);
+        long lastAcknowledged;
+        if (!acknowledgedMarkers.TryGetValue(controlAddress, out lastAcknowledged) || marker - lastAcknowledged > maxAckBatchSize) //ACK every N messages
```

Comment:
Probably want to use the session id here as well. Otherwise restarting the sender will not reset the ACKs on the receiver. This means the receiver won't send any acks until after it has been sent at least as many messages as it received in the previous session.